_posts/drafts/exploring-go-objects-part2.md

```diff
@@ -1,76 +1,411 @@
+It provides us with a compile time safe way to express protocols eliminating
+all the boilerplate of initializing structs with the proper functions.
+It will initialize the structs for us, and even optimize to
+initialize the struct just once, what is called in Go the **iface**,
+which is comparable to C++ **vtable**.
+
+It will also allow the code to be more decoupled since you don't have
+to known the package where the interface is defined to implement it.
+This is fundamental to how Go allows more flexibility with the same
+compile time safety as languages like Java/C++.
+
+Lets revisit the previous file protocol with interfaces:
 
 ```go
-type SomeProtocol interface {
-    Do(a string) error
-    DoOther(a int) (int, error)
+package main
+
+type Reader interface {
+    Read(data []byte) (int, error)
+}
+
+type Closer interface {
+    Close() error
+}
+
+type ReadCloser interface {
+    Reader
+    Closer
+}
+
+type Opener func() (ReadCloser, error)
+
+type File struct {}
+
+func (f *File) Read(data []byte) (int, error) {
+        return 0, nil
 }
 
-func useSomeProtocol(s SomeProtocol) {
-        s.Do("hi")
-        s.DoOther(5)
+func (f *File) Close() error {
+        return nil
+}
+
+func useFileProtocol(open Opener) {
+        f, _ := open()
+        data := make([]byte, 50)
+        f.Read(data)
+        f.Close()
 }
 
 func main() {
-        useSomeProtocol(SomeProtocol{
-            Do: func(a string) error {
-                    return nil
-            },
-            DoOther: func(a int) (int, error) {
-                    return 0, nil
-            },
+        useFileProtocol(func() (ReadCloser, error) {
+                return &File{}, nil
         })
 }
 ```
 
-The negative point of this is that there is no easy way to verify at
-compile time that all the functions of the interface instance
-have been initialized with a proper function.
+One key difference is that with interfaces this code is now
+safe. The **useFileProtocol** do not have to worry with calling
+functions that are actually nil, the Go compiler will do the work
+of creating a struct that holds a pointer with a **iface** descriptor
+that has all the functions required to satisfy the protocol.
+It does this per match of type <-> interface, as it is used (initializes
+on first usage). 
+
+You still can cause a segmentation fault if you do something like this:
+
+```go
+func() (ReadCloser, error) {
+        var a ReadCloser
+        return a, nil
+}
+```
+
+But there is still the bonus that every time you initialize the variable
+properly, like this :
+
+```go
+func() (ReadCloser, error) {
+        var a ReadCloser = &Whatever{}
+        return a, nil
+}
+```
+
+If it compiles you can be sure it is safe to call all functions of the interface.
+Also with the Go's interface mechanism an object can implement multiple different
+protocols without even knowing these protocols.
+
+How useful can be implementing protocols that you do not even known exists ?
```

Comment:
I think 'did not even know it exists' or 'do not even know it exists' sounds better :)

Reply:
agreed